Data/Aeson/Stream.hs

```diff
@@ -0,0 +1,272 @@
+{-# LANGUAGE BangPatterns, OverloadedStrings, RankNTypes #-}
+module Data.Aeson.Stream (
+    Token (..),
+    TokenStream,
+    tokenStream,
+    decodeWith,
+    eitherDecodeWith,
+    decodeValue,
+    parseValue,
+    ) where
+
+import Prelude ()
+import Prelude.Compat
+
+import Control.DeepSeq (NFData (..))
+import Control.Monad (ap)
+import Data.Aeson.Parser.Internal (jstring_, scientific)
+import Data.Aeson.Types.Internal (IResult(..), JSONPath, Result(..), Value(..))
+import Data.Scientific (Scientific)
+import Data.Text (Text)
+
+import qualified Data.Aeson.Types as Aeson
+import qualified Data.ByteString.Lazy as BSL
+import qualified Data.ByteString.Lazy.Char8 as BSL8
+import qualified Data.Attoparsec.ByteString.Lazy as A
+import qualified Data.Vector as V
+
+-- import Debug.Trace
+
+-------------------------------------------------------------------------------
+-- Token
+-------------------------------------------------------------------------------
+
+data Token
+    = TkNull
+    | TkTrue
+    | TkFalse
+    | TkArrayOpen
+    | TkArrayClose
+    | TkObjectOpen
+    | TkObjectClose
+    | TkKey !Text     -- ^ object keys use separate token
+    | TkText !Text
+    | TkNumber !Scientific
+    | TkError String  -- ^ error
+  deriving (Eq, Show)
+
+instance NFData Token where
+    rnf TkNull        = ()
+    rnf TkTrue        = ()
+    rnf TkFalse       = ()
+    rnf TkArrayOpen   = ()
+    rnf TkArrayClose  = ()
+    rnf TkObjectOpen  = ()
+    rnf TkObjectClose = ()
+    rnf (TkKey t)     = rnf t
+    rnf (TkText t)    = rnf t
+    rnf (TkNumber s)  = rnf s
+    rnf (TkError err) = rnf err
+
+type TokenStream = [Token]
+
+-------------------------------------------------------------------------------
+-- Token recognize
+-------------------------------------------------------------------------------
+
+newtype TokenParser = TokenParser
+    { runTokenParser :: [TokenParser] -> BSL.ByteString -> TokenStream }
```

Comment:
turning `[TokenParser]` into strict list (both elements and spine) doesn't seem to have effect on performance.